### PR TITLE
fix: add verb validations

### PR DESCRIPTION
**Pull request by sitaram-kalluri**
**- What I did**
1. Add key validations to the LLookupVerbBuilder and UpdateVerbBuilder
2. Update the documentation of the verb builders
3. Add the following unit tests to assert the validations
   * Tests to assert the changes in llookup_verb_builder.dart
      * Test to verify buildCommand method with local key
      * Test to verify buildCommand method with sharedWith key
      * Test to verify buildCommand method with public key
      * Test verify  buildCommand method with cached-sharedWith key
      * Test to verify buildCommand method with cached-public key
    *  Following are the negative tests
        * Test to verify cached local key throws invalid atkey exception
        * Test to verify local key with isPublic set to true throws invalid Atkey exception
        * Test to verify local key with sharedWith populated throws invalid Atkey exception
        * Test to verify isPublic set to true with sharedWith populated throws invalid Atkey exception
        * Test to verify Key cannot be null or empty
    * Tests to assert the changes in update_verb_builder.dart 
       * Test to verify local key with isPublic set to true throws invalid Atkey exception
       * Test to verify local key with sharedWith populated throws invalid Atkey exception
       * Test to verify isPublic set to true with sharedWith populated throws invalid Atkey exception
       * Test to verify Key cannot be null or empty

**- How I did it**

* Add validateKey method that validates the data in the verb builders
* Add buildKey method that returns the key from the verb builders

**- How to verify it**
* Unit tests are in place to assert the changes.

**- Description for the changelog**
* Add validations in the verb builders.

### CODE REVIEW

File: packages/at_commons/lib/src/keystore/at_key.dart [hunk 1]
```diff
@@ -48,6 +48,9 @@ class AtKey {
 
   set sharedWith(String? atSign) {
     assertStartsWithAtIfNotEmpty(atSign);
+    if(isLocal == true || metadata?.isPublic == true){
```
Comment (gkc) [inline packages/at_commons/lib/src/keystore/at_key.dart]: When the supplied atSign parameter is null, the checks for isLocal and isPublic should not be made.

Comment (gkc) [reply packages/at_commons/lib/src/keystore/at_key.dart]: Please write a test which first sets isLocal to true and then sets sharedWith.

File: packages/at_commons/lib/src/verb/update_verb_builder.dart [hunk 2]
```diff
@@ -274,4 +216,65 @@ class UpdateVerbBuilder implements VerbBuilder {
...
+    if (isPublic == true && sharedWith.isNotNull) {
+      throw InvalidAtKeyException(
+          'When isPublic is set to true, sharedWith cannot be populated');
+    }
+    if (atKey.isNull) {
+      throw InvalidAtKeyException('Key cannot be null or empty');
+    }
+  }
+
+  String buildKey() {
```
Comment (gkc) [inline packages/at_commons/lib/src/verb/update_verb_builder.dart]: This buildKey looks like a duplicate of the buildKey in llookup_verb_builder.dart

Please extract duplicate code into a single place.

**Review by gkc — CHANGES_REQUESTED** [review]

File: packages/at_commons/lib/src/verb/update_verb_builder.dart (hunk 2 above)
Comment (sitaram-kalluri) [reply packages/at_commons/lib/src/verb/update_verb_builder.dart]: Added abstract_verb_builder.dart to move the redundant code to a common place.

File: packages/at_commons/lib/src/keystore/at_key.dart (hunk 1 above)
Comment (sitaram-kalluri) [reply packages/at_commons/lib/src/keystore/at_key.dart]: @gkc: I am not clear on the first comment. If the comment is meant to skip the `IF` condition when atSign is null, then`assertStartsWithAtIfNotEmpty` method throws an exception if the atSign is null. So, the subsequent check will not execute.

Added an unit test where isLocal to set to true first and then sharedWith is added.

Comment (gkc) [reply packages/at_commons/lib/src/keystore/at_key.dart]: assertStartsWithAtIfNotEmpty will not throw an exception if the supplied value is null. so the subsequent check will indeed execute

File: packages/at_commons/lib/src/verb/abstract_verb_builder.dart [hunk 3]
```diff
@@ -24,7 +24,10 @@ abstract class AbstractVerbBuilder implements VerbBuilder {
     } else if (atKeyObj.sharedWith.isNotNull) {
       key += '${VerbUtil.formatAtSign(atKeyObj.sharedWith)}:';
     }
-    key += '${atKeyObj.key!}${VerbUtil.formatAtSign(atKeyObj.sharedBy)}';
+    key += atKeyObj.key!;
+    if (atKeyObj.sharedBy.isNotNull) {
```
Comment (gkc) [inline packages/at_commons/lib/src/verb/abstract_verb_builder.dart]: In what situations can we have sharedBy == null?

File: packages/at_commons/lib/src/keystore/at_key.dart (hunk 1 above)
Comment (gkc) [reply packages/at_commons/lib/src/keystore/at_key.dart]: please write a test where you are supplying a null sharedWith

**Review by gkc — CHANGES_REQUESTED** [review]

File: packages/at_commons/lib/src/keystore/at_key.dart (hunk 1 above)
Comment (sitaram-kalluri) [reply packages/at_commons/lib/src/keystore/at_key.dart]: Added a unit test and also refactored code when sharedWith is set to null on a local key or a public key

File: packages/at_commons/lib/src/verb/abstract_verb_builder.dart (hunk 3 above)
Comment (sitaram-kalluri) [reply packages/at_commons/lib/src/verb/abstract_verb_builder.dart]: These changes are removed.

**Comment by gkc** [comment]
Can you repeat the tests you did with AtmospherePro please, to make sure the changes all work? And rerun the at_client_sdk tests in your at_client_sdk PR also? And check on at_cli also.

**Comment by sitaram-kalluri** [comment]
> Can you repeat the tests you did with AtmospherePro please, to make sure the changes all work? And rerun the at_client_sdk tests in your at_client_sdk PR also? And check on at_cli also.

@gkc: Tested the changes with at_cli and atmosphere pro app. No issues were observed.

<details>
<summary>Test Results</summary>

at_cli test result:
----------------------
Able to update and llookup the key

```
sitaram@sitaram-ThinkPad-E14:~/IdeaProjects/atsign/core/at_tools/packages/at_cli$ dart run bin/main.dart -v update -a true -k firstname --value sitaram
3
sitaram@sitaram-ThinkPad-E14:~/IdeaProjects/atsign/core/at_tools/packages/at_cli$ dart run bin/main.dart -v llookup -a true -k firstname -b @sitaram
sitaram
sitaram@sitaram-ThinkPad-E14:~/IdeaProjects/atsign/core/at_tools/packages/at_cli$
```
atmosphere_pro app test results:
--------------------------------------------
1. Able to fetch the last received notification date time from the new key.
2. When the new key is unavailable, fetch the last received notification date time from the old key and update the new key.

update-3
</details>

**Comment by gkc** [comment]
Checks passed, closing PR

**Comment by gkc** [comment]
> Checks passed, closing PR

I meant - checks passed for [current trunk of at_client pointing to this branch for at_commons](https://github.com/atsign-foundation/at_client_sdk/actions/runs/3436988090)

File: packages/at_commons/lib/src/keystore/at_key.dart [hunk 4]
```diff
@@ -48,6 +49,10 @@ class AtKey {
 
   set sharedWith(String? atSign) {
     assertStartsWithAtIfNotEmpty(atSign);
+    if (atSign.isNotNull && (isLocal == true || metadata?.isPublic == true)) {
```
Comment (gkc) [inline packages/at_commons/lib/src/keystore/at_key.dart]: I have added the test case I asked for, see additions in at_key_test.dart in the commit I just made

File: packages/at_commons/lib/src/utils/string_utils.dart [hunk 5]
```diff
@@ -0,0 +1,13 @@
+/// Extending the String class to check null and empty.
+extension NullCheck on String? {
```
Comment (gkc) [inline packages/at_commons/lib/src/utils/string_utils.dart]: One last thing Sita

This extension should be called NullOrEmptyCheck and its methods should be called isNullOrEmpty and isNotNullOrEmpty

I have made these changes in the commit I just pushed to this branch

**Review by gkc — APPROVED** [review]
Nice work @kalluriramkumar - LGTM

I've made some changes to tidy up code a little for clarity and readability, and pushed a couple of commits. Please review them to make sure you are comfortable with the changes before merging.

File: packages/at_commons/lib/src/utils/string_utils.dart (hunk 5 above)
Comment (sitaram-kalluri) [reply packages/at_commons/lib/src/utils/string_utils.dart]: Thank you Gary.

File: packages/at_commons/lib/src/keystore/at_key.dart (hunk 4 above)
Comment (sitaram-kalluri) [reply packages/at_commons/lib/src/keystore/at_key.dart]: Got it Gary, Thank you.

**Pull request merged**